libs/x-content/src/main/java/org/elasticsearch/xcontent/support/filtering/FilterPath.java

```diff
@@ -99,6 +110,23 @@ boolean matches(String name, List<FilterPath> nextFilters) {
             nextFilters.add(this);
         }
 
+        // support dot in path
+        if (supportDotInFieldName && nextFilters.isEmpty()) {
```

Comment:
I haven't really wrapped my head around when nextFilters is empty. That's why I did mine in the caller. Could you explain what's up here?

Reply:
NextFilters are used to filter nested sub objects if current fieldName is matched and current FilterPathNode is not a final node. So if nextFilters is not empty, current fieldName must have been matched.

Reply:
And if nextFilters is not empty and our code reach here, current fieldName is not matched. We can then check whether it is the dot case or not.

Reply:
I see! So if `nextFilters` is non-empty that means that some other filter has already matched - it could be from a previous pattern.

Now, here's a question! Is it valid to skip? What if we had an object like `{"foo.bar": {"baz": 2}}` and two patterns `foo.bar.baz` and ` f*r.bort`. If the second one matches first we won't try to do the dots in field names thing and never notice the second one.

Now, my PR has similar but different problems for sure. If you turn on the mode it'll never match `f*r` for `foo.bar` keys. I'm realizing now this is wrong. Your approach here of running the test both times is good.